### PR TITLE
Working on tuning the avatar updates to take into account available per-node bandwidth as well as distance

### PR DESCRIPTION
**Pull request by jherico**
This CR is preparation for the task "Tune avatar mixer throttling for distance".  It restructures the AvatarMixer N^2 loop for readability and safety (the latter concerning the usage of mutex locking and unlocking) and adds inbound and outbound bandwidth recording to `NetworkPeer` from which `Node` is derived.  

I want to make sure this is the correct approach for the interface modification to `NetworkPeer`, even if the implementation using `BandwidthRecorder` is overkill (or possibly insufficient to our needs as the Asana task discusses using bandwidth calculated over a fixed historical time, while `BandwidthRecorder` does it over a fixed number of samples).

Another pull request will be created for updating the actual 'should I send node A's data to node B' calculation.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 72957670ba14c9130e0a3e42378bbfca18857ecb available at: https://gist.github.com/3c360e63e886f7cd83b4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4476.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4476.dmg

File: libraries/shared/src/TryLocker.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+#ifndef hifi_TryLocker_h
+#define hifi_TryLocker_h
+
+#include <QMutex>
+
+class MutexTryLocker {
```
Comment (ZappoMan) [inline libraries/shared/src/TryLocker.h]: coding standard: indents... looks like you've got 2 character indents on this file

**Comment by murillodigital** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 72957670ba14c9130e0a3e42378bbfca18857ecb available at: https://gist.github.com/3c360e63e886f7cd83b4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4476.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4476.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bf607e4c9cc634f6c383913c7be8ae2a1d340114 available at: https://gist.github.com/3c360e63e886f7cd83b4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4476.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4476.dmg

File: libraries/networking/src/NetworkPeer.cpp [hunk 2]
```diff
@@ -96,4 +97,38 @@ QDebug operator<<(QDebug debug, const NetworkPeer &peer) {
...
\ No newline at end of file
+}
+
+
+// FIXME this is a temporary implementation to determine if this is the right approach.
+// If so, migrate the BandwidthRecorder into the NetworkPeer class
+using BandwidthRecorderPtr = QSharedPointer < BandwidthRecorder > ;
+static QHash<QUuid, BandwidthRecorderPtr> PEER_BANDWIDTH;
+
+BandwidthRecorder & getBandwidthRecorder(const QUuid & uuid) {
```
Comment (ZappoMan) [inline libraries/networking/src/NetworkPeer.cpp]: coding standard: type & squish

File: libraries/networking/src/NetworkPeer.cpp [hunk 3]
```diff
@@ -96,4 +97,38 @@ QDebug operator<<(QDebug debug, const NetworkPeer &peer) {
...
+}
+
+
+void NetworkPeer::recordBytesSent(int count) {
+    auto & bw = getBandwidthRecorder(_uuid);
+    bw.updateOutboundData(0, count);
+}
+
+void NetworkPeer::recordBytesReceived(int count) {
+    auto & bw = getBandwidthRecorder(_uuid);
```
Comment (ZappoMan) [inline libraries/networking/src/NetworkPeer.cpp]: coding standard: type & squish

File: libraries/shared/src/TryLocker.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    QMutex & _mutex;
```
Comment (ZappoMan) [inline libraries/shared/src/TryLocker.h]: coding standard: type & squish

Comment (jherico) [inline libraries/shared/src/TryLocker.h]: fixed

File: libraries/shared/src/TryLocker.h (hunk 1 above)
Comment (jherico) [inline libraries/shared/src/TryLocker.h]: fixed

**Comment by hifi-gustavo** [comment]
Build console output for commit: 599fe09e75fe6b2148ef0a14b410be1aaa3b00c2 available at: https://gist.github.com/3c360e63e886f7cd83b4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4476.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4476.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7079b5f978514277f9eb65a939fb5d2f174f5206 available at: https://gist.github.com/3c360e63e886f7cd83b4<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4476.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4476.dmg

File: assignment-client/src/avatars/AvatarMixer.cpp [hunk 5]
```diff
@@ -132,83 +145,97 @@ void AvatarMixer::broadcastAvatarData() {
             
             AvatarData& avatar = nodeData->getAvatar();
             glm::vec3 myPosition = avatar.getPosition();
+            // TODO use this along with the distance in the calculation of whether to send an update 
+            // about a given otherNode to this node
+            // FIXME does this mean we should sort the othernodes by distance before iterating 
+            // over them?
+            float outputBandwidth = node->getOutboundBandwidth();
```
Comment (ZappoMan) [inline assignment-client/src/avatars/AvatarMixer.cpp]: note: this produces the following warning on XCode

```
/assignment-client/src/avatars/AvatarMixer.cpp:152:19: Unused variable 'outputBandwidth'
```

**Pull request merged**